jetstream/engine/aot_utils.py

```diff
@@ -0,0 +1,246 @@
+  prefill_buckets = [
+      bucket
+      for bucket in prefill_buckets
+      if bucket <= prefill_engine.max_prefill_length
+  ]
+  prefill_engine.prefill_buckets = prefill_buckets
+  if prefill_engine.max_prefill_length not in prefill_buckets:
+    prefill_buckets.append(prefill_engine.max_prefill_length)
+
+  def compile_prefill(length):
+    padded_tokens, true_length = jnp.ones((length), dtype="int32"), length
+
+    lowered = jax.jit(
+        prefill_engine._downstream_engine.prefill,  # pylint: disable=protected-access
+        out_shardings=prefill_engine.get_prefix_destination_sharding(),
+    ).lower(
+        params=prefill_params,
+        padded_tokens=padded_tokens,
+        true_length=true_length,
+    )
+    logging.info(
+        "---------Prefill engine %d lowered for prefill length %d.---------",
+        prefill_idx,
+        length,
+    )
+    compiled = lowered.compile()
+    logging.info(
+        "---------Prefill engine %d compiled for prefill length %d.---------",
+        prefill_idx,
+        length,
+    )
+    prefill_compiled[length] = compiled
+
+  logging.info("---------Prefill compilation %d begun.---------", prefill_idx)
+
+  prefill_compiled = {}
+  with concurrent.futures.ThreadPoolExecutor(
+      max_workers=len(prefill_buckets)
+  ) as executor:
+    _ = executor.map(compile_prefill, prefill_buckets)
+
+  prefill_engine.prefill_compiled = prefill_compiled
+
+  logging.info(
+      "---------Prefill compilation %d complete.---------", prefill_idx
+  )
+
+  return prefill_compiled
+
+
+def initialize_insert_generate_jit_cache(
+    *,
+    prefill_engine: engine_api.WarmedUpEngine,
+    generate_engine: engine_api.WarmedUpEngine,
+    prefill_params: Any,
+    generate_params: Any,
+    generate_idx: int,
+):
+  """Initialiszes jit cache for insert and generate.
+
+  Args:
+      generate_engine: A generate engine to be compiled for.
+      generate_params: The associated parameters.
+      generate_idx: Which generate engine it is.
+  """
+
+  prefill_buckets = token_utils.DEFAULT_PREFILL_BUCKETS
+  prefill_buckets = [
+      bucket
+      for bucket in prefill_buckets
+      if bucket <= generate_engine.max_prefill_length
+  ]
+  generate_engine.prefill_buckets = prefill_buckets
+  if generate_engine.max_prefill_length not in prefill_buckets:
+    prefill_buckets.append(generate_engine.max_prefill_length)
+
+  decode_state = generate_engine.init_decode_state()
+
+  def compile_insert(length):
+    padded_tokens, true_length = jnp.ones((length), dtype="int32"), length
+
+    prefill, _ = prefill_engine._downstream_engine.prefill(  # pylint: disable=protected-access
+        params=prefill_params,
+        padded_tokens=padded_tokens,
+        true_length=true_length,
+    )
+
+    lowered = jax.jit(generate_engine._downstream_engine.insert).lower(  # pylint: disable=protected-access
+        prefix=prefill, decode_state=decode_state, slot=1
+    )
```

Comment:
can you try printing out lowered.as_text() here? So we want to see if slot is hardcoded in the generated StableHLO or is captured as variable. If it's hardcoded then this line is going to recompile

Reply:
Discussed offline, slot is captured as variable